app/lib/payment_gateways/stripe_crypt.rb

```diff
@@ -15,35 +15,40 @@ def initialize(user)
     end
 
     def update!(params)
-      stripe_params = params.require(:stripe)
-      customer = create_stripe_customer(stripe_params)
-      update_account(customer, stripe_params)
-    rescue Stripe::CardError => error
-      report_error(error.message)
+      payment_method_id = params.require(:stripe).require(:payment_method_id)
+      payment_method = Stripe::PaymentMethod.retrieve(payment_method_id, api_key)
+      card = payment_method.card
+
+      account.credit_card_expires_on_month = card.exp_month
+      account.credit_card_expires_on_year = card.exp_year
+      account.credit_card_partial_number = card.last4
+      account.credit_card_auth_code = payment_method.customer
+      account.payment_method_id = payment_method_id
+      account.save
     end
 
-    def create_stripe_customer(stripe_params)
-      token   = stripe_params.require(:token)
-      api_key = payment_gateway_options.fetch(:login)
-      Stripe::Customer.create({ card: token,
-                                description: account.org_name,
-                                email: user.email,
-                                metadata: {
-                                  '3scale_account_reference' => buyer_reference
-                                }
-                              }, api_key)
+    def create_stripe_customer
+      Stripe::Customer.create({
+        description: account.org_name,
+        email: user.email,
+        metadata: {  '3scale_account_reference' => buyer_reference }
+      }, api_key)
     end
 
-    def update_account(customer, stripe_params)
-      account.credit_card_expires_on_month = stripe_params.require(:expires_on_month)
-      account.credit_card_expires_on_year = stripe_params.require(:expires_on_year)
-      account.credit_card_partial_number = stripe_params.require(:partial_number)[-4..-1]
-      account.credit_card_auth_code = customer.id
-      account.save
+    def create_stripe_setup_intent(customer = create_stripe_customer)
+      Stripe::SetupIntent.create({
```

Comment:
Are we using the stripe gem here now instead of activemerchant?? I see the activemerchant gem doesn't support the setupintents API and only the [paymentintents API](https://github.com/activemerchant/active_merchant/blob/abc50bf93aff4dac8368119fd81280fc7f1e07d4/lib/active_merchant/billing/gateways/stripe_payment_intents.rb)

Reply:
No, the activemerchant gem supports setup intents, and in fact, we are using them. Check  this:

https://github.com/activemerchant/active_merchant/blob/abc50bf93aff4dac8368119fd81280fc7f1e07d4/lib/active_merchant/billing/gateways/stripe_payment_intents.rb#L7

In this implementation, we are using Stripe gem to create the customer/setupIntent, and the activemerchant gem to 'charge' the customer. This is only because it is the way how we were doing it before and the less we touch in a hurry, the less risk of breaking things and having to spend Christmas fixing them (or worse).
But FYI nowadays we might be able to implement all of this with activemerchant only without the stripe gem.